src/cache_cf.cc

```diff
@@ -311,139 +312,219 @@ skip_ws(const char* s)
+    RejectTrailingGarbage("integer", raw, numberParser);
+    return result;
 }
 
+/// IsIfStatementOpening() helper that interprets input prefix as a preprocessor condition
 static bool
-StrToInt(const char* str, long& number)
+EvalBoolExpr(Parser::Tokenizer &tk)
 {
-    assert(str != nullptr);
+    const auto operand = ExtractOperand("preprocessor condition", tk);
+
+    static const auto keywordTrue = SBuf("true");
+    if (operand == keywordTrue)
+        return true;
 
-    char* end;
-    number = strtol(str, &end, 0);
+    static const auto keywordFalse = SBuf("false");
+    if (operand == keywordFalse)
+        return false;
+
+    const auto lhs = operand;
+
+    const auto op = ExtractOperator("equality sign in an equality condition", tk);
+    static const auto keywordEqual = SBuf("=");
+    if (op != keywordEqual)
+        throw TextException(ToSBuf("expected equality sign (=) but got ", op), Here());
+
+    const auto rhs = ExtractOperand("right-hand operand of an equality condition", tk);
+    return EvalNumber(lhs) == EvalNumber(rhs);
+}
+
+/// interprets input as the first line of a preprocessor `if` statement
+/// \returns std::nullopt if input does not look like an `if` statement
+/// \returns `if` condition value if input is an `if` statement
+static std::optional<bool>
+IsIfStatementOpening(Parser::Tokenizer tk)
+{
+    // grammar: space* "if" space condition space* END
+    (void)SkipOptionalSpace(tk);
+    const auto keywordIf = SBuf("if");
+    if (tk.skip(keywordIf) && SkipOptionalSpace(tk)) {
+        const auto condition = tk.remaining();
+        const auto result = EvalBoolExpr(tk);
+        (void)SkipOptionalSpace(tk);
+        RejectTrailingGarbage("preprocessor condition", condition, tk);
+        return result;
+    }
 
-    return (end != str) && (*end == '\0'); // returns true if string contains nothing except number
+    // e.g., "iffy_error_responses on"
+    return std::nullopt;
 }
 
+/// interprets input as an `else` or `endif` line of a preprocessor `if` statement
+/// \returns false if input does not look like an `else` or `endif` line
 static bool
-EvalBoolExpr(const char* expr)
+IsIfStatementLine(const SBuf &keyword, Parser::Tokenizer tk)
 {
-    assert(expr != nullptr);
-    if (strcmp(expr, "true") == 0) {
-        return true;
-    } else if (strcmp(expr, "false") == 0) {
-        return false;
-    } else if (const char* equation = strchr(expr, '=')) {
-        const char* rvalue = skip_ws(equation + 1);
-        char* lvalue = (char*)xmalloc(equation - expr + 1);
-        xstrncpy(lvalue, expr, equation - expr + 1);
-        trim_trailing_ws(lvalue);
+    // grammar: space* keyword space* END
+    (void)SkipOptionalSpace(tk);
+    if (tk.skip(keyword)) {
+        if (tk.atEnd())
+            return true;
+
+        if (SkipOptionalSpace(tk)) {
+            RejectTrailingGarbage("preprocessor keyword", keyword, tk);
+            return true;
+        }
+        // e.g., "elseif"
```

Comment:
Ah, "elseif" is not supported.

However, despite `cf.data.pre` stating that "NOTE: An else-if condition is not supported."  the configuration grammar:
```
   // grammar: "else" space "if" space condition space* END
```
does actually work in current Squid. This new logic will break that. Please fix.